handlers/common.go

```diff
@@ -116,3 +141,121 @@ func initEnviron(machineDir string) []string {
 	}
 	return env
 }
+
+func createExtractedConfig(event *events.Event, machine *client.Machine) (string, error) {
+	// We are going to ignore doing anything for VirtualBox given that there is no way you can
+	// use Machine once it has been created.  Virtual is mainly a test-only use case
+	if strings.ToLower(machine.Driver) == "virtualbox" {
+		log.Debug("VirtualBox machine does not need config extracted")
+		return "", nil
+	}
+
+	// We will now zip, base64 encode the machine directory created, and upload this to cattle server.  This can be used to either recover
+	// the machine directory or used by Rancher users for their own local machine setup.
+	log.WithFields(log.Fields{
+		"resourceId": event.ResourceId,
+	}).Info("Creating and uploading extracted machine config")
+
+	// tar.gz the $CATTLE_HOME/machine/<machine-id>/machines/<machine-name> dir (v0.2.0)
+	// Open the source directory and read all the files we want to tar.gz
+	baseDir, err := getBaseMachineDir(machine.ExternalId)
+	if err != nil {
+		return "", err
+	}
+
+	machineDir := filepath.Join(baseDir, "machines", machine.Name)
+	dir, err := os.Open(machineDir)
+	if err != nil {
+		return "", err
+	}
+	defer dir.Close()
+
+	log.WithFields(log.Fields{
+		"machineDir": machineDir,
+	}).Debug("Preparing directory to be tar.gz")
+
+	// Be able to read the files under this dir
+	files, err := dir.Readdir(0)
+	if err != nil {
+		return "", err
+	}
+
+	// create the tar.gz file
+	destFile := filepath.Join(baseDir, machine.Name+".tar.gz")
+	tarfile, err := os.Create(destFile)
+	if err != nil {
+		return "", err
+	}
+
+	defer tarfile.Close()
+	fileWriter := gzip.NewWriter(tarfile)
+	defer fileWriter.Close()
+
+	tarfileWriter := tar.NewWriter(fileWriter)
+	defer tarfileWriter.Close()
+
+	// Read and add files into <machine-name>.tar.gz
+	for _, fileInfo := range files {
+
+		// For now, we will skip directories.  If we need to zip directories, we need to revist this code
+		if fileInfo.IsDir() {
+			continue
+		}
+
+		file, err := os.Open(filepath.Join(machineDir, fileInfo.Name()))
+		if err != nil {
+			return "", err
+		}
+
+		defer file.Close()
+
+		// prepare the tar header
+		header := new(tar.Header)
+		header.Name = filepath.Join(machine.Name, fileInfo.Name())
+		header.Size = fileInfo.Size()
+		header.Mode = int64(fileInfo.Mode())
+		header.ModTime = fileInfo.ModTime()
+
+		err = tarfileWriter.WriteHeader(header)
+		if err != nil {
+			return "", err
+		}
+
+		_, err = io.Copy(tarfileWriter, file)
+		if err != nil {
+			return "", err
+		}
+	}
+	return destFile, nil
+}
+
+func getIdempotentExtractedConfig(machine *client.Machine, machineDir string, apiClient *client.RancherClient) (string, error) {
+	if machine.Driver == "virtualbox" {
```

Comment:
Can we extract this to a common function as the previous check in createExtractedConfig.